examples/deep_learning/distributed_flax_imagenet.py

```diff
@@ -0,0 +1,839 @@
+        ResNet, stage_sizes=[3, 4, 6, 3], block_cls=ResNetBlock),
+    'ResNet50': functools.partial(
+        ResNet, stage_sizes=[3, 4, 6, 3], block_cls=BottleneckResNetBlock),
+    'ResNet101': functools.partial(
+        ResNet, stage_sizes=[3, 4, 23, 3], block_cls=BottleneckResNetBlock),
+    'ResNet152': functools.partial(
+        ResNet, stage_sizes=[3, 8, 36, 3], block_cls=BottleneckResNetBlock),
+    'ResNet200': functools.partial(
+        ResNet, stage_sizes=[3, 24, 36, 3], block_cls=BottleneckResNetBlock),
+}
+
+
+### SPMD utilities.
+
+
+def setup_data_parallel_mesh():
+  global_mesh = Mesh(np.asarray(jax.devices(), dtype=object), ['data'])
+  jax.experimental.maps.thread_resources.env = (
+      jax.experimental.maps.ResourceEnv(physical_mesh=global_mesh, loops=()))
+
+
+### Training loop utilities.
+
+
+class LossFnAux(NamedTuple):
+  """Stores auxiliary values computed during loss function eval for reuse."""
+  batch_stats: ArrayTree
+  logits: Array
+
+
+def cross_entropy_loss(labels, logits, **_):
+  xentropy = jax.vmap(jaxopt_losses.multiclass_logistic_loss)(labels, logits)
+  return jnp.mean(xentropy)
+
+
+def loss_fun(
+    params: ArrayTree,
+    batch_stats: ArrayTree,
+    batch: Batch,
+    model: nn.Module,
+    weight_decay: float = 1e-4,
+) -> Tuple[Numeric, LossFnAux]:
+  """Loss function used for training."""
+  logits, new_mutable_variables = model.apply(
+      {'params': params, 'batch_stats': batch_stats},
+      batch['image'],
+      mutable=['batch_stats'])
+
+  xentropy = cross_entropy_loss(labels=batch['label'], logits=logits)
+  weight_penalty_params = [x for x in jax.tree_leaves(params) if x.ndim > 1]
+  weight_l2 = tree_util.tree_l2_norm(weight_penalty_params, squared=True)
+  loss = xentropy + weight_decay * 0.5 * weight_l2
+
+  new_batch_stats = new_mutable_variables['batch_stats']
+  aux = LossFnAux(batch_stats=new_batch_stats, logits=logits)
+
+  return loss, aux
+
+
+@struct.dataclass
+class TrainMetrics(clu_metrics.Collection):
+
+  accuracy: clu_metrics.Accuracy
+  learning_rate: clu_metrics.Average.from_output('learning_rate')
+  loss: clu_metrics.Average.from_output('loss')
+  xent: clu_metrics.Average.from_fun(cross_entropy_loss)
+
+
+@struct.dataclass
+class EvalMetrics(clu_metrics.Collection):
+
+  accuracy: clu_metrics.Accuracy
+  xent: clu_metrics.Average.from_fun(cross_entropy_loss)
+
+
+def train_step(
+    params: ArrayTree,
+    state: ArrayTree,
+    batch: Batch,
+    metrics: Optional[TrainMetrics],
+    learning_rate_fn: LearningRateFn,
+    solver: jaxopt.OptaxSolver,
+) -> Tuple[ArrayTree, ArrayTree, TrainMetrics]:
+  """Performs a single training step."""
+  # Retrieves Flax mutables from previous step, stored in `state.aux`.
+  batch_stats = state.aux.batch_stats
+  # Computes updated model parameters and optimizer state.
+  params, state = solver.update(
+      params=params,
+      state=state,
+      batch_stats=batch_stats,
+      batch=batch,
```

Comment:
In `run_iterator`, we call it `data`. Maybe `batch` would have been a better name in retrospect, as it conveys the idea of "subsample"? Since not so many people use `run_iterator` yet, I believe, maybe it's not too late to change? I believe these small naming consistencies are very important to provide a nice user experience.

Reply:
Yes, that's a good point. I "inherited"  `batch` from the Flax ImageNet example, but certainly +1 for consistency with `run_iterator`. I have no strong preference myself on whether to use `data` or `batch`.